archive/m/modula2/FractionMath.mod

```diff
@@ -0,0 +1,213 @@
+MODULE FractionMath;
+
+FROM StrIO IMPORT WriteString, WriteLn;
+
+TRACED CLASS Fraction;
+    REVEAL GreatestCommonDivisor, Reduce, SetFraction, ShowFraction, [=], [>], [<], [>=], [<=],
+           [+], [-], [*], [/]; 
+    VAR 
+    numerator: INTEGER;
+    denominator: INTEGER;
+    gcd : INTEGER;
+
+    PROCEDURE SetFraction(n, d : INTEGER);
+    BEGIN
+        numerator := n;
+        denominator := d;
+    END SetFraction;
+
+    PROCEDURE GreatestCommonDivisor(n, d : INTEGER) : INTEGER;
+    BEGIN
+        IF d = 0 THEN
+            WriteString("error cannot divide by 0");
+        END;
+
+        WHILE n # d DO 
+            IF n > d THEN 
+                n := n - d;
+            ELSE
+                d := d - n;
+            END;
+        END;
+    
+        RETURN n;
+    END GreatestCommonDivisor;
+
+    PROCEDURE Reduce(n, d : INTEGER);
+    BEGIN 
+        IF d = 0 THEN
+            WriteString("error cannot divide by 0");
+        
+        ELSE
+            gcd := GreatestCommonDenominator(n, d);
+            numerator := n / gcd;
+            denominator := d / gcd;
+        END;
+    END Reduce; 
```

Comment:
Seems like this procedure is never used. There is a lot of code duplication where this could have been used. I'd recommend just making this part of `SetFraction`